packages/integration-boilerplate-node/src/auth.ts

```diff
@@ -0,0 +1,163 @@
+import { Base64 } from "js-base64";
+import jwksRsa from "jwks-rsa";
+import jwt from "jsonwebtoken";
+import fetch from "node-fetch";
+import { Logger } from "winston";
+import { URLSearchParams } from "url";
+import NodeCache from "node-cache";
+
+import { PermissionDeniedError } from "./errors";
+
+function createSaasquatchJwksClient(appDomain: string) {
+  return jwksRsa({
+    jwksUri: `https://${appDomain}/.well-known/jwks.json`,
+    cache: true,
+    cacheMaxEntries: 5, // We only have 2 right now
+  });
+}
+
+/**
+ * Validate the given JWT with public JWKS and get the claims.
+ * @param token The input JWT
+ */
+function validateWithJwks(
+  token: string,
+  jwksClient: jwksRsa.JwksClient,
+  options?: jwt.VerifyOptions
+): Promise<object> {
+  return new Promise((resolve, reject) => {
+    jwt.verify(
+      token,
+      (header, callback) => {
+        jwksClient.getSigningKey(header.kid!, (err, key) => {
+          callback(err, key ? key.getPublicKey() : undefined);
+        });
+      },
+      options,
+      (err, decoded) => {
+        if (err) {
+          reject(err);
+        } else {
+          if (!decoded) {
+            reject();
+          } else {
+            resolve(decoded);
+          }
+        }
+      }
+    );
+  });
+}
+
+export class Auth {
+  private jwksClientCache: NodeCache;
+  private saasquatchTokenCache: NodeCache;
+
+  constructor(
+    private appDomain: string,
+    private auth0ClientId: string,
+    private auth0ClientSecret: string,
+    private auth0Domain: string,
+    private logger: Logger
+  ) {
+    this.jwksClientCache = new NodeCache({
+      stdTTL: 86400, // 1 day
+      useClones: false,
+      maxKeys: 1,
+    });
+    this.saasquatchTokenCache = new NodeCache({
+      stdTTL: 3600 * 6, // 6 hours
+      maxKeys: 1,
+    });
+  }
+
+  private getJwksClient() {
```

Comment:
After refactoring the FSR integration to use the integration-boilerplate-node package, I encountered an issue when handling many webhooks at near the same time:
```
info: POST /webhook
info: POST /webhook
info: POST /webhook
info: POST /webhook
error: {
  errorCode: 'PERMISSION_DENIED',
  error: 'Cache max keys amount exceeded'
}
error: {
  errorCode: 'PERMISSION_DENIED',
  error: 'Cache max keys amount exceeded'
}
error: {
  errorCode: 'PERMISSION_DENIED',
  error: 'Cache max keys amount exceeded'
}

```
This only occurs on the first set of webhooks received. What I believe is happening is that `getJwksClient` is called multiple times, and the check `this.jwksClientCache.has("the")` returns false, and too many clients are attempted to be created in the NodeCache.

This isn't a critical issue as the webhook processing fails early, and will be retried. It could probably be fixed with something along the lines of:
```
if (this.jwksClientCache.has("the")) {
  this.logger.debug(`Retrieving JWKS client from from cache`);
  return this.jwksClientCache.get("the") as jwksRsa.JwksClient;
}

const client = createSaasquatchJwksClient(this.appDomain);
try {
  this.jwksClientCache.set("the", client);
} catch (e) {
  if (this.jwksClientCache.has("the")) {
    this.logger.debug(`Retrieving JWKS client from from cache`);
    return this.jwksClientCache.get("the") as jwksRsa.JwksClient;
  }
  throw e;
}
return client;
 
```

Reply:
Huh, thanks @ianhitchcock , this is really interesting. I'll see if I can replicate this locally because it really shouldn't happen. I would expect in a single-threaded, event loop driven application that the cache is not eventually consistent, so perhaps I misunderstood the `node-cache` API somewhere.